lib/Core/Executor.cpp

```diff
@@ -1544,36 +1560,56 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
 #else
           BasicBlock *caseSuccessor = si->getSuccessor(i);
 #endif
-          std::map<BasicBlock*, ref<Expr> >::iterator it =
-            targets.insert(std::make_pair(caseSuccessor,
-                           ConstantExpr::alloc(0, Expr::Bool))).first;
 
-          it->second = OrExpr::create(match, it->second);
+          // Handle the case that a basic block might be the target of multiple
+          // switch cases
+          // Currently we generate an expression containing all switch-case
+          // values for the same target basic block. We spare us forking too
+          // many times but we generate more complex condition expressions
+          // TODO make allow use to choose the forking option too
+          std::pair<std::map<BasicBlock *, ref<Expr> >::iterator, bool> res =
+              branchTargets.insert(std::make_pair(
+                  caseSuccessor, ConstantExpr::alloc(0, Expr::Bool)));
+
+          res.first->second = OrExpr::create(match, res.first->second);
+          if (res.second) {
+            bbOrder.push_back(caseSuccessor);
```

Comment:
@MartinNowack I think it would be good to put a comment here explaining the guard because it would be very easy to introduce a subtle bug here. As you state in the comments you are handling the case where multiple cases target the same basic block. However ``bbOrder`` isn't a set like data structure which means without the ``if (res.second)`` guard we might insert the same ``BasicBlock`` multiple times (this would result in creating multiple ``ExecutionState`` which are identical and therefore redundant).